EIPS/eip-signTypedData.md

```diff
@@ -0,0 +1,439 @@
+[trezor]: https://github.com/trezor/trezor-mcu/issues/163
+
+Both determinism and invectiveness would be trivially true if `len(b)` was left out entirely. The point is, it is difficult to map arbitrary sets to bytestrings without introducing security issues in the encoding function. Yet the current design of `eth_sign` still takes a bytestring as input and expects implementors to come up with an encoding.
+
+### Messages
+
+The `eth_sign` call assumes messages to be bytestrings. In practice we are not hashing bytestrings but the collection of all semantically different messages of all different DApps `𝕄`. Unfortunately, this set is impossible to formalize  so we approximate it with the set of typed named structures `𝕊` and a domain separator `𝔹²⁵⁶` to obtain the set `𝔹²⁵⁶ × 𝕊`. This standard formalizes the set `𝕊` and provides a deterministic injective encoding function for `𝔹²⁵⁶ × 𝕊`.
+
+
+## Specification
+
+<!-- The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (cpp-ethereum, go-ethereum, parity, ethereumj, ethereumjs, ...).  -->
+
+The set of signable messages is extended from transactions and bytestrings `𝕋 ∪ 𝔹⁸ⁿ` to also include structured data `𝕊`. The new set of signable messages is thus `𝕋 ∪ 𝔹⁸ⁿ ∪ 𝕊`. They are encoded to bytestrings suitable for hashing and signing as follows:
+
+* `encode(t : 𝕋) = RLP_encode(t)`
+* `encode(b : 𝔹⁸ⁿ) = "\x19Ethereum Signed Message:\n" ‖ len(b) ‖ b` where `len(b)` is the *non-zero-padded* ascii-decimal encoding of the number of bytes in `b`.
+* `encode((d, s) : 𝔹²⁵⁶ × 𝕊) = "\x01" ‖ d ‖ hashStruct(s)` where `d` is a domain separator and `hashStruct(s)` is defined below.
+
+This encoding is deterministic because the individual components are. The encoding is injective because the three cases always differ in first byte. (`RLP_encode(t)` does not start with `\x01` or `\x19`.)
+
+### Domain separator `d`
+
+The domain separator is a 256-bit nonce. For each distinct use-case a new nonce is generated. It is recommended to use the address of a relevant contract or a randomly generated 256-bit constant. It should be considered part of the interface specification and shared with it.
+
+It is recommended to use a different domain separator on different chains (i.e. testnet and mainnet) to prevent accidental compatibility. The Ethereum signature method already includes a chain-id to prevent cross-chain signature validity.
+
+### Definition of typed structured data `𝕊`
+
+To define the set of all structured data, we start with defining acceptable types. Like ABIv2 these are closely related to Solidity types. It is illustrative to adopt Solidity notation to explain the definitions. The standard is specific to the Ethereum Virtual Machine, but aims to be agnostic to higher level languages. Example:
+
+```cpp
+struct Message {
```

Comment:
It'd be worth explicitly describing the format used here as a language, so a tool could convert it into a typeHash independent of Solidity.

Reply:
Well, it's a pseudo-language I use to define the language used for `encodeType`. IMHO, the `encodeType` language itself is already suitable for human consumption, the derivation from a Solidity-ish pseudo-language serves just to explain it further. Formalizing the pseudo-language would just move the problem one meta-level up :)

I will instead specify the `encodeType` language more formally (BNF grammar, etc) and use that to explain the language.

(I also considered taking the small subset of Solidity relating to struct declaration and using that directly. The advantage would be that an `encodeType` can be directly used in Solidity source code, which makes it trivial to keep in sync. Downside is this would place a number of weird constraints on the Solidity side to make it unambiguous. It did not seem worth it, plus the ABI already set a precedent of using a DSL.)